torch/distributed/fsdp/flat_param.py

```diff
@@ -575,6 +586,404 @@ def shard_metadata(
+
+    def unshard(self):
+        """
+        Runs the unshard logic. This includes all-gathering the flattened
+        parameter and switching to using the unsharded flattened parameter. If
+        the handle does not need unsharding, then this only switches to using
+        the unsharded flattened parameter. For ``NO_SHARD``, this is a no-op.
+
+        If FSDP is in :meth:`summon_full_params` and the handle uses parameter
+        mixed precision, then the parameter is forced to full precision.
+        """
+        if not self.needs_unshard():
+            if self.uses_sharded_strategy:
+                # The handle may have been resharded without freeing the padded
+                # unsharded flattened parameter, in which case we need to
+                # switch to using the unsharded parameter
+                unsharded_flat_param = self._get_padded_unsharded_flat_param()
+                self._use_unsharded_flat_param(unsharded_flat_param)
+            return
+        unsharded_flat_param = self._alloc_padded_unsharded_flat_param()
+        self._all_gather_flat_param(unsharded_flat_param)
+
+    def needs_unshard(self) -> bool:
+        """Returns if the handle's flattened parameter needs to be unsharded."""
+        if not self.uses_sharded_strategy:
+            return False
+        unsharded_flat_param = self._get_padded_unsharded_flat_param()
+        already_unsharded = (
+            unsharded_flat_param.storage().size() == unsharded_flat_param.numel()
+        )
+        return not already_unsharded
+
+    def _alloc_padded_unsharded_flat_param(self):
+        """
+        Allocates the *padded* unsharded flattened parameter. The unpadded
+        unsharded flattened parameter is always a view into the padded one.
+        This padded parameter is saved to a different attribute on the
+        ``FlatParameter`` depending on if we force full precision.
+        """
+        self._check_sharded_strategy()
+        flat_param = self.flat_param
+        padded_unsharded_size = flat_param._full_param_padded.size()  # type: ignore[attr-defined]
+        unsharded_flat_param = self._get_padded_unsharded_flat_param()
+        self._check_storage_freed(unsharded_flat_param)
+        _alloc_storage(unsharded_flat_param, padded_unsharded_size)
+        return unsharded_flat_param
+
+    def _get_padded_unsharded_flat_param(self) -> torch.Tensor:
+        """
+        Returns a reference to the padded unsharded flattened parameter
+        depending on the calling context. This should only be called if using a
+        sharded strategy.
+        """
+        self._check_sharded_strategy()
+        flat_param = self.flat_param
+        if self._force_full_precision:
+            # When parameter mixed precision is enabled, we use a different
+            # tensor as the all-gather destination to preserve the invariant
+            # that  `_full_param_padded` is in the low precision
```

Comment:
wondering why we need to keep this invariant, it is empty anyway. could we just use `_full_param_padded` for full precision data even in '_force_full_precision' mode?

Reply:
Context: Internally, we always preserve the Python `Tensor` variable (in this case, `_full_param_padded`), which has the **correct metadata** -- notably, the `dtype`, `shape`, and strides. We manually allocate and free the `Tensor` variable by calling `resize_()`, which only takes the number of bytes as argument. The `Tensor` variable's metadata determines how to interpret the underlying storage's bytes.

If we broke the invariant and used `_full_param_padded` for the full precision too, then we cannot directly call `_full_param_padded.storage().resize_(padded_unsharded_size)`. We would have to modify `_full_param_padded`'s metadata to use the full precision and remember to restore the metadata later. This does not fit with our current allocation/free strategy of only calling `resize_()`.

Aside: This actually differs from DeepSpeed's implementation, which does not preserve the `Tensor` variable and instead lets the garbage collection free the tensor naturally.
https://github.com/microsoft/DeepSpeed/blob/b288cf1b9b90030cc2025f2681e77f3bc81dd409/deepspeed/runtime/zero/partition_parameters.py#L237-L239
By setting `param.data = torch.empty(0, dtype=param.dtype, device=param.device)`, the original `param.data` and its storage have 0 refcount and can be garbage collected.

Reply:
I thought we want to use storage() to preserve the original data ptr, e.g., after free and resizing the full parameters to be 0 in the forward, it can all gather and resize back to the same storage data ptr for the full parameters in the backward pass. This is important, because autograd holds a reference or pointer of this full parameter variable during forward and uses this reference or pointer in the backward pass. 

But I may be wrong...., because I feel the resizing is quite fragile to make sure storage ptr to be the same in the forward and backward

Reply:
I do not think that is the case. As a sanity check, if after the `resize_(0)` to free, the `resize_(<actual size>)` to allocate was able to keep the same data pointer, then that means no other object can use the contiguous memory that was freed. This is equivalent to not freeing the storage at all.

In my understanding, autograd only needs to hold the reference to the **`Tensor` variable**, in this case the `FlatParameter: nn.Parameter`, not the data pointer. When we call `flat_param.storage().resize_(0)` to free, that free is invisible to autograd (and pretty much everything else). In fact, all of the metadata about the `flat_param` is the same, as if its storage were still allocated. In that sense, FSDP is "playing with fire". It can free using `storage().resize_(0)` as long as it calls `storage().resize_(<actual size>)` before the tensor is involved in any computation (e.g. the backward gradient computation). If it does not, then the execution will error with a message related to the storage.

---
Here is an example:
```
>>> import torch
>>> t1 = torch.randn((3, 5))
>>> t1.data_ptr()
94152269896640
>>> t1.storage().data_ptr()
94152269896640
>>> t1.storage().resize_(0)  # free
>>> t1.data_ptr()
0
>>> t1.storage().data_ptr()
0
>>> t1.shape
torch.Size([3, 5])
>>> t1.storage().resize_(15)  # reallocate
>>> t1.data_ptr()
94152206816576
>>> t1.storage().data_ptr()
94152206816576
```
We see that the data pointer is not preserved (and also that `t.storage().data_ptr()` is equivalent to `t.data_ptr()`).

---
Actually, the storage object changes after `resize_()` too.
```
>>> t2 = torch.randn((3, 5))
>>> id(t2.storage())
140082535361504
>>> t2.storage().resize_(0)
>>> id(t2.storage())
140082535362176
>>> t2.storage().resize_(15)
>>> id(t2.storage())
140082535361648
```

Reply:
thanks for explanation and confirmation, that actually relieved my concern:)